### PR TITLE
add test for auth and merge auth to etcd client

### PR DESCRIPTION
**Pull request by StupidHod**
Look forward your review, @xiang90 @AdoHe .

### CODE REVIEW

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 1]
```diff
@@ -33,4 +67,46 @@ public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder bui
...
+
+    /**
+     * get token from etcd with name and password
+     *
+     * @param channel channel to etcd
+     * @param name auth name
+     * @param password auth password
+     * @return token string
+     */
+    private String getToken(ManagedChannel channel, ByteString name, ByteString password) {
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: authenticate instead of getToken?

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 2]
```diff
@@ -33,4 +67,46 @@ public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder bui
...
+        try {
+            ListenableFuture<AuthenticateResponse> authResp = AuthGrpc.newFutureStub(channel)
+                    .authenticate(
+                            AuthenticateRequest.newBuilder()
+                                    .setNameBytes(name)
+                                    .setPasswordBytes(password)
+                                    .build());
+            return authResp.get().getToken();
+        }catch (InterruptedException ie){
+            return null;
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: we probably should throw out this exception. user should know the auth fails.

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 3]
```diff
@@ -33,4 +67,46 @@ public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder bui
...
+    /**
+     * get token from etcd with name and password
+     *
+     * @param channel channel to etcd
+     * @param name auth name
+     * @param password auth password
+     * @return token string
+     */
+    private String getToken(ManagedChannel channel, ByteString name, ByteString password) {
+
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: remove the empty line

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 4]
```diff
@@ -14,15 +23,40 @@
...
         this.channel = this.channelBuilder.build();
 
-        KVGrpc.KVFutureStub kvStub = KVGrpc.newFutureStub(this.channel);
+        kvStub = KVGrpc.newFutureStub(this.channel);
+        authStub = AuthGrpc.newFutureStub(this.channel);
+
+        String token=null;
+
+        if(builder.getName()!=null && builder.getPassword()!=null){
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: throw exception if one of them is empty and one of them is not.

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 5]
```diff
@@ -14,15 +23,40 @@
...
+            token = getToken(this.channel, builder.getName(), builder.getPassword());
+        }
+
+        if(token!=null) {
+            /**
+             * add token to channel header
+             */
+            Metadata metadata = new Metadata();
+            Metadata.Key<String> TOKEN_KEY= Metadata.Key.of("token", Metadata.ASCII_STRING_MARSHALLER);
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: in go, there is a utility func to set creds: https://github.com/coreos/etcd/blob/edcfcae332679cf02396d8a0f743bed047a06768/clientv3/client.go#L202

Is there a similar thing in java?

File: src/main/java/com/coreos/jetcd/EtcdClient.java (hunk 1 above)
Comment (StupidHod) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: OK.

File: src/main/java/com/coreos/jetcd/EtcdClient.java (hunk 2 above)
Comment (StupidHod) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: Hi, should we pack this exception with AuthFailedException, or just throw the origin exception?

File: src/main/java/com/coreos/jetcd/EtcdClient.java (hunk 4 above)
Comment (StupidHod) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: Ok.

File: src/main/java/com/coreos/jetcd/EtcdClient.java (hunk 2 above)
Comment (AdoHe) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: what we expect user to do if we throw this exception?

Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: password might be wrong. user should fill in the correct one. connection might be bad, need to retry.

File: src/main/java/com/coreos/jetcd/EtcdClient.java (hunk 5 above)
Comment (AdoHe) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: @StupidHod you may take a look at `withCallCredentials` method of stub.

Comment (StupidHod) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: Hi, MetadataUtils.attachHeaders is a tool method to add metadata, which is short and easy to use.
MetadataUtils.attachHeaders has do the same thing as withCallCredentials.
`            authStub.withCallCredentials(new CallCredentials() {
                @Override
                public void applyRequestMetadata(MethodDescriptor<?, ?> methodDescriptor, Attributes attributes, Executor executor, MetadataApplier metadataApplier) {
                    metadataApplier.apply(metadata);
                }
            });`

Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: let's use withCallCredential. That is more "official" way to do this sort of thing

**Comment by mitake** [comment]
@xiang90 sure, could you wait until this weekend? For my private reasons I cannot work on it immediately.

**Comment by xiang90** [comment]
@mitake sure thanks.

File: src/test/java/com/coreos/jetcd/EtcdAuthClientTest.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+     * @throws InterruptedException
+     */
+    @BeforeTest
+    public void setupEnv() throws AuthFailedException, ConnectException {
+        test = new Assertion();
+        etcdClient =
+                EtcdClientBuilder.newBuilder().endpoints("localhost:2379").build();
+        kvClient = etcdClient.getKVClient();
+        authClient = etcdClient.getAuthClient();
+
```
Comment (xiang90) [inline src/test/java/com/coreos/jetcd/EtcdAuthClientTest.java]: remove empty line.

File: src/test/java/com/coreos/jetcd/EtcdAuthClientTest.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+     * @throws ExecutionException
+     * @throws InterruptedException
+     */
+    @Test(dependsOnMethods = "testEnableAuth", groups = "authEnable")
+    public void setupAuthClient() throws AuthFailedException, ConnectException {
+        etcdClient =
+                EtcdClientBuilder.newBuilder().endpoints("localhost:2379").setName(userName).setPassword(password).build();
+        kvClient = etcdClient.getKVClient();
+        authClient = etcdClient.getAuthClient();
+
```
Comment (xiang90) [inline src/test/java/com/coreos/jetcd/EtcdAuthClientTest.java]: remove empty line.

**Comment by xiang90** [comment]
@StupidHod Can you fix the CI?

File: src/test/java/com/coreos/jetcd/EtcdAuthClientTest.java (hunk 7 above)
Comment (xiang90) [inline src/test/java/com/coreos/jetcd/EtcdAuthClientTest.java]: remove empty line.

**Comment by xiang90** [comment]
@StupidHod Have you tested the auth path? Enable auth, add a role A with rw to key foo, add a user A, assign role A to User A. Without user A and its password, client cannot write to foo. With user A and its correct password, client can write to foo.

**Comment by mitake** [comment]
@xiang90 @StupidHod I'm looking around this PR and want to try it. Does jetcd have an working example?

**Comment by StupidHod** [comment]
@xiang90 Ok, I will try to fix the CI and add more tests.

**Comment by StupidHod** [comment]
@mitake You can clone from my repository and switch to auth branch to run this test.

File: scripts/run_etcd.sh [hunk 8]
```diff
@@ -1,5 +1,10 @@
 #!/bin/bash
-
+git clone https://go.googlesource.com/go
+cd go
+git checkout master
+cd src
+./all.bash
+GOROOT_BOOTSTRAP=$GOROOT
```
Comment (hongchaodeng) [inline scripts/run_etcd.sh]: Can we download etcd release instead of building it everytime?

Comment (AdoHe) [inline scripts/run_etcd.sh]: @StupidHod @hongchaodeng actually in PR #30 I have update this, ptal.

Comment (StupidHod) [inline scripts/run_etcd.sh]: @hongchaodeng You advice is good.

Comment (StupidHod) [inline scripts/run_etcd.sh]: @AdoHe Your version seems better, you can overwrite my version latter, as I want to pass the test. =,=

**Comment by hongchaodeng** [comment]
Hi @StupidHod , Can you try to use an IDE that formats java files :)
It would be much easier to read code like that. For example, the if-else block should have more spaces.

**Comment by StupidHod** [comment]
@hongchaodeng OK, tks for your advice.

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 9]
```diff
@@ -1,28 +1,70 @@
...
+
+        String token=null;
+
+        if(builder.getName()!=null && builder.getPassword()!=null){
+            token = authenticate(this.channel, builder.getName(), builder.getPassword());
+        }else if(!(builder.getName()==null&&builder.getPassword()==null))
+        {
+            throw new NullPointerException("Username or password can not be null.");
+        }
+        if(token!=null) {
```
Comment (hongchaodeng) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: Can you change the code path to:
```
if case1:
  throw e1
else if case2:
  throw e2

token = authenticate(...)
// use token below...
```

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 10]
```diff
@@ -1,28 +1,70 @@
...
+            /**
+             * add token to channel header
+             */
+            Metadata metadata = new Metadata();
+            Metadata.Key<String> TOKEN_KEY= Metadata.Key.of(TOKEN, Metadata.ASCII_STRING_MARSHALLER);
```
Comment (hongchaodeng) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: This place has too many details and is confusing..
Can you help make them a function with a comprehensive function name?

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 11]
```diff
@@ -33,4 +75,47 @@ public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder bui
...
+     * @param name auth name
+     * @param password auth password
+     * @return token string
+     * @throws  ConnectException This may be caused as network reason, wrong address
+     * @throws  AuthFailedException This may be caused as wrong username or password
+     */
+    private String authenticate(ManagedChannel channel, ByteString name, ByteString password) throws ConnectException, AuthFailedException {
+        try {
+            ListenableFuture<AuthenticateResponse> authResp = AuthGrpc.newFutureStub(channel)
+                    .authenticate(
```
Comment (hongchaodeng) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: move `authenticate` to previous line

File: src/main/java/com/coreos/jetcd/EtcdClient.java [hunk 12]
```diff
@@ -33,4 +75,47 @@ public EtcdClient(ManagedChannelBuilder<?> channelBuilder, EtcdClientBuilder bui
...
+     * get token from etcd with name and password
+     *
+     * @param channel channel to etcd
+     * @param name auth name
+     * @param password auth password
+     * @return token string
+     * @throws  ConnectException This may be caused as network reason, wrong address
+     * @throws  AuthFailedException This may be caused as wrong username or password
+     */
+    private String authenticate(ManagedChannel channel, ByteString name, ByteString password) throws ConnectException, AuthFailedException {
```
Comment (hongchaodeng) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: Can we split this function into two? One is authenticate returning auth response; the other is get token (maybe we don't even need second func).

File: src/main/java/com/coreos/jetcd/EtcdClientBuilder.java [hunk 13]
```diff
@@ -54,14 +59,48 @@ public EtcdClientBuilder endpoints(String ... endpoints) {
...
+    }
+
+    /**
+     * config etcd auth name
+     *
+     * @param name etcd auth name
+     * @return this builder
+     * @throws NullPointerException if name is null
+     */
+    public EtcdClientBuilder setName(ByteString name) {
```
Comment (hongchaodeng) [inline src/main/java/com/coreos/jetcd/EtcdClientBuilder.java]: why we are using ByteString?

File: src/main/java/com/coreos/jetcd/EtcdClient.java (hunk 9 above)
Comment (StupidHod) [inline src/main/java/com/coreos/jetcd/EtcdClient.java]: OK.

File: src/main/java/com/coreos/jetcd/EtcdClientBuilder.java (hunk 13 above)
Comment (StupidHod) [inline src/main/java/com/coreos/jetcd/EtcdClientBuilder.java]: @hongchaodeng , We use this as this is the string used by gRPC which support many encoding.

**Comment by mitake** [comment]
@StupidHod I mean I want some examples that can be executed with running etcd servers. Or the test can be executed? Sorry, I don't know well about testing stuff of Java...

File: scripts/run_etcd.sh [hunk 14]
```diff
@@ -3,10 +3,8 @@
...
-cd etcd
-echo "Building etcd..."
-./build
-echo "Done!"
+curl -L https://github.com/coreos/etcd/releases/download/v3.0.4/etcd-v3.0.4-linux-amd64.tar.gz -o etcd-v3.0.4-linux-amd64.tar.gz
+tar xzvf etcd-v3.0.4-linux-amd64.tar.gz && cd etcd-v3.0.4-linux-amd64
+./etcd --version
 
-nohup bash -c 'bin/etcd &'
+nohup bash -c './etcd &'
```
Comment (hongchaodeng) [inline scripts/run_etcd.sh]: have a new line

File: src/test/java/com/coreos/jetcd/EtcdClientBuilderTest.java [hunk 15]
```diff
@@ -33,7 +35,7 @@ public void testEndPoints_Verify_SomeEmpty() {
     }
 
     @Test(expectedExceptions = IllegalStateException.class)
-    public void testBuild_WithoutEndpoints() {
+    public void testBuild_WithoutEndpoints() throws AuthFailedException, ConnectException {
         builder.build();
     }
 }
```
Comment (hongchaodeng) [inline src/test/java/com/coreos/jetcd/EtcdClientBuilderTest.java]: have a new line at the end of the file

**Comment by hongchaodeng** [comment]
@mitake 
I think we (probably @StupidHod ) can make some examples separately. Let's move forward with this PR.
LGTM after fixing the nits.

**Comment by StupidHod** [comment]
@hongchaodeng Well, I will make some examples late, sorry for the nits issues, I have fixt them.

**Comment by hongchaodeng** [comment]
LGTM.
Thanks @StupidHod for the work.

**Comment by mitake** [comment]
@hongchaodeng @StupidHod I see, I'd like to check the behavior of the code with an working example later.

**Pull request merged**